### ISSUE TITLE
Orc does not call a custom destructor

### ISSUE THREAD

**Issue by Clyybber** [nim-lang/Nim#18421]
The cycle created by this snippet never get's collected, as evidenced by the `=destroy` hook never getting called.

### Example
```nim
type
  ThingObj = object
    name: string
    
  Thing = ref ThingObj

  C = ref object of RootObj
    result: Thing
    child: C

proc `=destroy`(t: var ThingObj) =
  echo "destroyed ", t.name

proc main =
  var c = C()
  c.result = Thing(name: "leak one")
  c.child = C(child: c)

main()
GC_fullCollect()
```

### Current Output
nothing

### Expected Output
```
destroyed leak one
```

```
$ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-07-02
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 60eac79f7f0478565a24ff423aa52ab53527d153
active boot switches: -d:release -d:danger
```

**Comment by GordonBGood** [comment]
I shorted the main to as follows so one can just comment out the last line to see it destroys as normal for non-cyclic, issue is as you describe when there is a cycle; also `getOccupiedMem()` indicates the memory is recovered, just that the destructor for the internal object field is not run:
```nim
proc main =
  var c = C(result: Thing(name: "leak one"))
  c.child = c

let mem = getOccupiedMem()
main()
GC_fullCollect()
echo "Memory balance:  ", getOccupiedMem() - mem
```
So this seems to indicate that your problem here isn't so much a leak as a bug not calling the destructor, but that of course could potentially cause a leak if one was dependent on the destructor to do further memory cleanups.  In this case, if the string isn't a string constant then I guess its pointer into the heap wouldn't be cleaned up but that doesn't show up on the memory measurement because it isn't "managed" memory.

Trying it with the @araq's newish `-d:nimThinout` also doesn't work (Memory balance:  80) as per my issue #18420, so that seems to indicate that the issue there is more than just for nested cyclic closures as I show there.

**EDIT_ADD:**  It occurs to me that this "not calling object destructors on objects for which cycles are detected" may be the the manifestation of the same root compiler bug for my issue as per the above link, as it would also explain what happens there.

**Comment by Araq** [comment]
`-d:nimThinout`  is a failed experiment, don't use it.

**Comment by Araq** [comment]
You need to use

```nim


proc `=destroy`(t: var ThingObj) =
  echo "destroyed ", t.name
  `=destroy`(t.name)

```

in order to prevent leaks btw.

**Comment by GordonBGood** [comment]
I can't help fix this issue as it is something that Orc is defaulting to in the compiler magic, but I can perhaps help with the quality control test procedure to make sure Orc has a proper acceptance test, as follows:

1. I got sidetracked into trying to use `-d:nimThinout` by seeing that my Hamming Numbers Test was using that, so assumed that it was a production feature.
2. Since it is not a production feature as per @araq's comment above, then it means that the test is not being used as a default compiler acceptance test by testament?
3. It should likely be a testament standard acceptance test for Orc (it can't be used with Arc because it contains cyclic references).
4. I can modify it back to its original purpose, tune it so it only does checking using `doAssert`'s, and add slightly to it so it checks that the number of references created are the same as the number destroyed.
5. If you like the idea, I can post the code here so you can include it in the PR that fixes this issue.

The Hamming Numbers Test is likely the most thorough test for Orc that can be had as it contains cyclic `ref`'s that can span many and variable numbers of chained structures but some references are cyclic and some may not be as base values are consumed.  It is thus more thorough than the short linear chain as per the code in the issue I referenced above.

I played with it a bit just now and it indicates that when finding the first 20 Hamming numbers with the current implementation of Orc, 52 lazy list references are created but only the ones that are destroyed by the iterator in doing copies are properly destroyed as in 36, leaving 16 not properly destroyed even though the references are force recovered by the Orc cycle detector.  I can easily make it check that the difference is zero after a `GC_FullCollect()`.

I could also write something simpler not containing cyclic references to check Arc.  Something like a Fibonacci Lazy List using linear primitives should do the trick.

**Comment by Araq** [comment]
> I played with it a bit just now and it indicates that when finding the first 20 Hamming numbers with the current implementation of Orc, 52 lazy list references are created but only the ones that are destroyed by the iterator in doing copies are properly destroyed as in 36, leaving 16 not properly destroyed even though the references are force recovered by the Orc cycle detector. I can easily make it check that the difference is zero after a GC_FullCollect().

Yeah, please submit the test.

> I could also write something simpler not containing cyclic references to check Arc. Something like a Fibonacci Lazy List using linear primitives should do the trick.

I don't know how important that is.

Believe it or not, --gc:orc was heavily stress tested with examples carefully crafted to make it fail. Apparently your tests are the better ones though. ;-)

**Comment by GordonBGood** [comment]
@araq:

> Believe it or not, --gc:orc was heavily stress tested with examples carefully crafted to make it fail. Apparently your tests are the better ones though. ;-)

I'm sorry if I gave the impression that I thought that Orc hadn't been tested properly - obviously, it has, else it wouldn't work as well as it does.  But there can never be enough tests, obviously - if we could anticipate every kind of error, then there wouldn't be any need for tests at all as our code would be perfect.  All I meant is that perhaps my Hamming Number Test may take a slightly different approach than what has already been done.  I think it likely that this bug is just forgetting to couple the forced destruction that Orc does with our "gcDestructors" technique.

This new code has been corrected slightly as there was an error in calculating the conversion for `BigInt`'s larger than 2^32 and I've added some extra metrics; Code as follows (Using it with the `-d:trace20` compiler option may help find this problem):

Note:  using "trace20", It may be helpful to note that the values that aren't being destroyed are both the large values (inside their ref objects) above the limit that haven't been used (and thus consumed and force destroyed), but also the captured cyclic values (in their ref objects) that are referenced by the capture chain.  I suppose one could modify the program even further to show the capture nesting "tree", but I haven't done that.  Also note that a given value (other than "one" which is only used three times as the bases for all subsequent values) will appear at least twice: once when it is created by the given multiplication, and then perhaps many times as it is captured by subsequent "merge" operations.  That may be why this test should be so thorough:  the "cycle span" increases with increasing hamming values, but also there are multiple embedded capture cycles.  There is also that each node represents an integer value which is fairly easy to recognize as to patterns.

```nim
# test Nim Hamming Number Lazy List algo with reference counts and not...
# compile with "-d:release -d:danger" and test with various
# memory managment GC's, allocators, threading, etc.
# it should be guaranteed to work with zero memory leaks with `--gc:orc`...

# compile with `-d:trace20` to trace creation and destruction of first 20 values.

from math import log2

# implement our own basic BigInt so the bigints library isn't necessary...
type
  BigInt = object
    digits: seq[uint32]
let zeroBigInt = BigInt(digits: @[ 0'u32 ])
let oneBigInt = BigInt(digits: @[ 1'u32 ])

proc shladd(bi: var BigInt; n: int; a: BigInt) =
  # assume that both `bi` and `a` are sized correctly with
  # msuint32 for both not containing a zero
  let alen = a.digits.len
  let mx = max(bi.digits.len, a.digits.len)
  for i in bi.digits.len ..< mx: bi.digits.add 0'u32
  var cry = 0'u64
  for i in 0 ..< alen:
    cry += (bi.digits[i].uint64 shl n) + a.digits[i].uint64
    bi.digits[i] = cry.uint32; cry = cry shr 32
  for i in alen ..< mx:
    cry += bi.digits[i].uint64 shl n
    bi.digits[i] = cry.uint32; cry = cry shr 32
  if cry > 0'u64:
    bi.digits.add cry.uint32

proc `$`(x: BigInt): string =
  if x.digits.len == 0 or (x.digits.len == 1 and x.digits[0] == 0'u32):
    return "0"
  result = ""; var n = x; var msd = n.digits.high
  while msd >= 0:
    if n.digits[msd] == 0'u32: msd.dec; continue
    var brw = 0.uint64
    for i in countdown(msd, 0):
      let dvdnd = n.digits[i].uint64 + (brw shl 32)
      let q = dvdnd div 10'u64; brw = dvdnd - q * 10'u64
      n.digits[i] = q.uint32
    result &= $brw
  for i in 0 .. result.high shr 1: # reverse result string in place
    let tmp = result[^(i + 1)]
    result[^(i + 1)] = result[i]
    result[i] = tmp

type TriVal = (uint32, uint32, uint32)
type LogRep = (float64, TriVal)
type LogRepf = proc(x: LogRep): LogRep
const one: LogRep = (0.0'f64, (0'u32, 0'u32, 0'u32))
proc `<`(me: LogRep, othr: LogRep): bool = me[0] < othr[0]

proc convertTriVal2BigInt(tpl: TriVal): BigInt =
  result = oneBigInt
  let (x2, x3, x5) = tpl
  for _ in 1 .. x2: result.shladd 1, zeroBigInt
  for _ in 1 .. x3: result.shladd 1, result
  for _ in 1 .. x5: result.shladd 2, result

const lb2 = 1.0'f64
const lb3 = 3.0'f64.log2
const lb5 = 5.0'f64.log2

proc mul2(me: LogRep): LogRep =
  let (lr, tpl) = me; let (x2, x3, x5) = tpl
  (lr + lb2, (x2 + 1, x3, x5))

proc mul3(me: LogRep): LogRep =
  let (lr, tpl) = me; let (x2, x3, x5) = tpl
  (lr + lb3, (x2, x3 + 1, x5))

proc mul5(me: LogRep): LogRep =
  let (lr, tpl) = me; let (x2, x3, x5) = tpl
  (lr + lb5, (x2, x3, x5 + 1))

type
  LazyListObj = object
    hd: LogRep
    tlf: proc(): LazyList {.closure.}
    tl: LazyList
  LazyList = ref LazyListObj

var dstcnt = 0
proc `=destroy`(ll: var LazyListObj) =
  dstcnt += 1
  if ll.tlf == nil and ll.tl == nil: return
  when defined(trace20):
    echo "destroying:  ", (dstcnt, ll.hd[1].convertTriVal2BigInt)
  if ll.tlf != nil: ll.tlf.`=destroy`
  if ll.tl != nil: ll.tl.`=destroy`
  wasmoved(ll)

proc rest(ll: LazyList): LazyList = # not thread-safe; needs lock on thunk
  if ll.tlf != nil: ll.tl = ll.tlf(); ll.tlf = nil
  ll.tl

var crtcnt = 0
iterator hammings(until: int): TriVal =
  proc merge(x, y: LazyList): LazyList =
    let xh = x.hd; let yh = y.hd; crtcnt += 1
    when defined(trace20):
      echo "merge create:  ", (crtcnt - 1, (if xh < yh: xh else: yh)[1].convertTriVal2BigInt)
    if xh < yh: LazyList(hd: xh, tlf: proc(): auto = merge x.rest, y)
    else: LazyList(hd: yh, tlf: proc(): auto = merge x, y.rest)
  proc smult(mltf: LogRepf; s: LazyList): LazyList =
    proc smults(ss: LazyList): LazyList =
      when defined(trace20):
        echo "mult create:  ", (crtcnt, ss.hd.mltf[1].convertTriVal2BigInt)
      crtcnt += 1; LazyList(hd: ss.hd.mltf, tlf: proc(): auto = ss.rest.smults)
    s.smults
  proc unnsm(s: LazyList, mltf: LogRepf): LazyList =
    var r: LazyList = nil
    when defined(trace20):
      echo "first create:  ", (crtcnt, one[1].convertTriVal2BigInt)
    let frst = LazyList(hd: one, tlf: proc(): LazyList = r); crtcnt += 1
    r = if s == nil: smult(mltf, frst) else: s.merge smult(mltf, frst)
    r
  yield one[1]
  var hmpll: LazyList = ((nil.unnsm mul5).unnsm mul3).unnsm mul2
  for _ in 2 .. until:
    yield hmpll.hd[1]; hmpll = hmpll.rest # almost forever

proc main =
  var s = ""
  for h in hammings(20): s &= $h.convertTrival2BigInt & " "
  doAssert s == "1 2 3 4 5 6 8 9 10 12 15 16 18 20 24 25 27 30 32 36 ",
           "Algorithmic error finding first 20 Hamming numbers!!!"

  when not defined(trace20):
    var lsth: TriVal; crtcnt = 0; dstcnt = 0
    for h in hammings(200): lsth = h
    doAssert $lsth.convertTriVal2BigInt == "16200",
             "Algorithmic error finding 200th Hamming number!!!"

  when not declared(thinout): GC_FullCollect()

let mem = getOccupiedMem()
main()
let mb = getOccupiedMem() - mem
doAssert mb == 0, "Found memory leak of " & $mb & " bytes!!!"
doAssert crtcnt == dstcnt,
         "Missing calls to destructors!!! Created:  " &
         $crtcnt & "; Destroyed:  " & $dstcnt
```

**Comment by Araq** [comment]
Again, Orc does not leak. The destructor is not called but that is a compiler bug, not an Orc bug. See for yourself:

```nim

discard """
  output: ''''''
"""

# bug #18421

# test Nim Hamming Number Lazy List algo with reference counts and not...
# compile with "-d:release -d:danger" and test with various
# memory managment GC's, allocators, threading, etc.
# it should be guaranteed to work with zero memory leaks with `--gc:orc`...

# compile with `-d:trace20` to trace creation and destruction of first 20 values.

from math import log2

# implement our own basic BigInt so the bigints library isn't necessary...
type
  BigInt = object
    digits: seq[uint32]
let zeroBigInt = BigInt(digits: @[ 0'u32 ])
let oneBigInt = BigInt(digits: @[ 1'u32 ])

proc shladd(bi: var BigInt; n: int; a: BigInt) =
  # assume that both `bi` and `a` are sized correctly with
  # msuint32 for both not containing a zero
  let alen = a.digits.len
  let mx = max(bi.digits.len, a.digits.len)
  for i in bi.digits.len ..< mx: bi.digits.add 0'u32
  var cry = 0'u64
  for i in 0 ..< alen:
    cry += (bi.digits[i].uint64 shl n) + a.digits[i].uint64
    bi.digits[i] = cry.uint32; cry = cry shr 32
  for i in alen ..< mx:
    cry += bi.digits[i].uint64 shl n
    bi.digits[i] = cry.uint32; cry = cry shr 32
  if cry > 0'u64:
    bi.digits.add cry.uint32

proc `$`(x: BigInt): string =
  if x.digits.len == 0 or (x.digits.len == 1 and x.digits[0] == 0'u32):
    return "0"
  result = ""; var n = x; var msd = n.digits.high
  while msd >= 0:
    if n.digits[msd] == 0'u32: msd.dec; continue
    var brw = 0.uint64
    for i in countdown(msd, 0):
      let dvdnd = n.digits[i].uint64 + (brw shl 32)
      let q = dvdnd div 10'u64; brw = dvdnd - q * 10'u64
      n.digits[i] = q.uint32
    result &= $brw
  for i in 0 .. result.high shr 1: # reverse result string in place
    let tmp = result[^(i + 1)]
    result[^(i + 1)] = result[i]
    result[i] = tmp

type TriVal = (uint32, uint32, uint32)
type LogRep = (float64, TriVal)
type LogRepf = proc(x: LogRep): LogRep
const one: LogRep = (0.0'f64, (0'u32, 0'u32, 0'u32))
proc `<`(me: LogRep, othr: LogRep): bool = me[0] < othr[0]

proc convertTriVal2BigInt(tpl: TriVal): BigInt =
  result = oneBigInt
  let (x2, x3, x5) = tpl
  for _ in 1 .. x2: result.shladd 1, zeroBigInt
  for _ in 1 .. x3: result.shladd 1, result
  for _ in 1 .. x5: result.shladd 2, result

const lb2 = 1.0'f64
const lb3 = 3.0'f64.log2
const lb5 = 5.0'f64.log2

proc mul2(me: LogRep): LogRep =
  let (lr, tpl) = me; let (x2, x3, x5) = tpl
  (lr + lb2, (x2 + 1, x3, x5))

proc mul3(me: LogRep): LogRep =
  let (lr, tpl) = me; let (x2, x3, x5) = tpl
  (lr + lb3, (x2, x3 + 1, x5))

proc mul5(me: LogRep): LogRep =
  let (lr, tpl) = me; let (x2, x3, x5) = tpl
  (lr + lb5, (x2, x3, x5 + 1))

type
  LazyListObj = object
    hd: LogRep
    tlf: proc(): LazyList {.closure.}
    tl: LazyList
  LazyList = ref LazyListObj

var dstcnt = 0
proc `=destroy`(ll: var LazyListObj) =
  dstcnt += 1
  if ll.tlf == nil and ll.tl == nil: return
  when defined(trace20):
    echo "destroying:  ", (dstcnt, ll.hd[1].convertTriVal2BigInt)
  if ll.tlf != nil: ll.tlf.`=destroy`
  if ll.tl != nil: ll.tl.`=destroy`
  wasmoved(ll)

proc rest(ll: LazyList): LazyList = # not thread-safe; needs lock on thunk
  if ll.tlf != nil: ll.tl = ll.tlf(); ll.tlf = nil
  ll.tl

var crtcnt = 0
iterator hammings(until: int): TriVal =
  proc merge(x, y: LazyList): LazyList =
    let xh = x.hd; let yh = y.hd; crtcnt += 1
    when defined(trace20):
      echo "merge create:  ", (crtcnt - 1, (if xh < yh: xh else: yh)[1].convertTriVal2BigInt)
    if xh < yh: LazyList(hd: xh, tlf: proc(): auto = merge x.rest, y)
    else: LazyList(hd: yh, tlf: proc(): auto = merge x, y.rest)
  proc smult(mltf: LogRepf; s: LazyList): LazyList =
    proc smults(ss: LazyList): LazyList =
      when defined(trace20):
        echo "mult create:  ", (crtcnt, ss.hd.mltf[1].convertTriVal2BigInt)
      crtcnt += 1; LazyList(hd: ss.hd.mltf, tlf: proc(): auto = ss.rest.smults)
    s.smults
  proc unnsm(s: LazyList, mltf: LogRepf): LazyList =
    var r: LazyList = nil
    when defined(trace20):
      echo "first create:  ", (crtcnt, one[1].convertTriVal2BigInt)
    let frst = LazyList(hd: one, tlf: proc(): LazyList = r); crtcnt += 1
    r = if s == nil: smult(mltf, frst) else: s.merge smult(mltf, frst)
    r
  yield one[1]
  var hmpll: LazyList = ((nil.unnsm mul5).unnsm mul3).unnsm mul2
  for _ in 2 .. until:
    yield hmpll.hd[1]; hmpll = hmpll.rest # almost forever

proc main =
  var s = ""
  for h in hammings(20): s &= $h.convertTrival2BigInt & " "
  doAssert s == "1 2 3 4 5 6 8 9 10 12 15 16 18 20 24 25 27 30 32 36 ",
           "Algorithmic error finding first 20 Hamming numbers!!!"

  when not defined(trace20):
    var lsth: TriVal; crtcnt = 0; dstcnt = 0
    for h in hammings(200): lsth = h
    doAssert $lsth.convertTriVal2BigInt == "16200",
             "Algorithmic error finding 200th Hamming number!!!"

let mem = getOccupiedMem()
main()
GC_FullCollect() # <----- do this call outside of `main`!
let mb = getOccupiedMem() - mem
doAssert mb == 0, "Found memory leak of " & $mb & " bytes!!!"
doAssert crtcnt == dstcnt,
         "Missing calls to destructors!!! Created:  " &
         $crtcnt & "; Destroyed:  " & $dstcnt
```

**Comment by GordonBGood** [comment]
@araq:

> Again, Orc does not leak. The destructor is not called but that is a compiler bug, not an Orc bug. See for yourself:

Yes, I now understand that Arc and Orc do not leak and that this issue is correctly titled as to what the problem really is:  that some destructors are not being called when `GC_FullCollect()` is run;  None of these tests prove one way or another wither the destructors are being called when whatever memory quota//limit that triggers an automatic collection is reached or not as these tests may not be big enough?

That's why the modified Hamming Numbers Test keeps track of when a custom destructor is called compared to when the object is constructed to see if there is a pattern that would show you where the problem is.

To me, since the problem only occurs when using `--gc:orc`, it appears as an Orc bug even though it isn't a memory leak.  However, it could cause a memory leak if a program is relying on a custom destructor to dealloc a pointer, for instance.

Then, after the problem is fixed, if it is run as a QA test, it shows that no future changes cause a regression, at least in these areas.

**Comment by Araq** [comment]
This works:

```nim

type
  ThingObj = object
    name: string
    
  Thing = ref ThingObj

  C = ref object of RootObj
    result: Thing
    child: C

proc `=destroy`(t: var ThingObj) =
  echo "destroyed ", t.name
  `=destroy`(t.name)

proc `=dispose`(t: var ThingObj) =
  echo "destroyed ", t.name
  `=destroy`(t.name)

proc main =
  var c = C()
  c.result = Thing(name: "leak one")
  c.child = C(child: c)

main()
GC_fullCollect()
```

There is a subtle difference between `=destroy` and `=dispose` that we need to document well. However, a mutable `=trace` would also help ... it's an open design question. Currently `=dispose` is not documented at all.

**Comment by GordonBGood** [comment]
Astounding!  So there is no bug, just a lack in documentation, which need I see has been elevated to **Serious** as it should be, as this is particularly important for maintainers and library writers.  From the repo code, it seems that there is also a possible implementation of `=dispose` for `--gc:arc`.

If the following code is added to the Hamming Numbers Test just above the `=destroy` proc, everything balances for `--gc:orc`:
```nim
proc `=dispose`(ll: var LazyListObj) =
  dstcnt += 1
  if ll.tlf == nil and ll.tl == nil: return
  when defined(trace20):
    echo "disposing:  ", (dstcnt, ll.hd[1].convertTriVal2BigInt)
  if ll.tlf != nil: ll.tlf.`=destroy`
```
So everything is fine for `--gc:orc` and `--gc:arc` (when it applies when there aren't cyclic references) once this is documented.

However, that seems to bring up a few interesting questions:

1. Is disposal guaranteed when using all of the garbage collectors that came before these new ones?  It seems they have a perhaps limited notion of these "hook" based proc's, and when the Hamming Number Test is run it shows an imbalance between the number of creations and destructions (including disposals) for the cyclic references used for all of the older garbage collectors.
2. Does this mean that manual allocations inside those objects won't be guaranteed to be deallocated by the object custom `=destroy`/`=dispose` unless there is some sort of alternate `ref object` finalizer used (which option does, in fact, exist in the system library for `new`)?
3. Although use of conventional garbage collection may eventually be depreciated (hopefully), shouldn't this requirement be well documented in the meantime if there is a possibility that not calling destructors may leak manually allocated heap memory that is expected to be cleaned up by the destructors?

**Comment by Araq** [comment]
@GordonBGood with https://github.com/nim-lang/Nim/pull/18440 `=dispose` will be gone and we can finally document `=trace`, it works as I envisioned it to work.

**Comment by GordonBGood** [comment]
@Araq:

I'll wait for your commit to be merged and check it out.  I hope it documents the use of `=trace` as to when it might be necessary  to use a custom version and how that custom version would be implemented.

**Comment by GordonBGood** [comment]
@Araq:

I see the commit, compiled latest devel as of today, the 9th of July 2021, and tested it.  **It runs a treat** (beautifully)!

I see you've added my "Hamming Numbers Test" to the "orc" tests, slightly modified as follows:

1. Obviously, you've eliminated the `=dispose` hook, which is no longer needed.
2. At the end of the `=destroy` hook, you eliminated the `wasmoved` at the end; I assume that the system takes care of marking the object as already destroyed so recursive destructions aren't run, and take instruction from this.
3. **The test program is not quite correct** when run without the `-d:trace20` option as I zeroed the `created` and `destroyed` counters between the two Hamming sequences, which leaves these counters out of sync as the new "Orc" has already destroyed some elements (my bad).  The correct code should be something like the following:
```nim
proc main =
  var s = ""
  for h in hammings(20): s &= $h.convertTrival2BigInt & " "
  doAssert s == "1 2 3 4 5 6 8 9 10 12 15 16 18 20 24 25 27 30 32 36 ",
           "Algorithmic error finding first 20 Hamming numbers!!!"

  when not defined(trace20):
    var lsth: TriVal # don't touch those created/destroyed counters!
    for h in hammings(200): lsth = h
    doAssert $lsth.convertTriVal2BigInt == "16200",
             "Algorithmic error finding 200th Hamming number!!!"

when defined(gcOrc):
  let mem = getOccupiedMem()
  main()
  GC_FullCollect()
  let mb = getOccupiedMem() - mem
  doAssert mb == 0, "Found memory leak of " & $mb & " bytes!!!"
  doAssert created == destroyed,
          "Missing calls to destructors!!! Created:  " &
          $created & "; Destroyed:  " & $destroyed
  when defined(nimAllocStats): echo getAllocStats()
else: main()
```
I have no idea of the purpose of a custom `=trace`, as the code base does't seem to have any examples of its use, but I assume that further documentation of this is coming.  **I do note that the proc signature for `=trace` of `proc[T: object](o: var T) {...}` is not quite enforced** as the compiler seems to also accept a non-`var` parameter.

**Comment by Araq** [comment]
Feel free to create a PR that changes the test for the better.

**Comment by GordonBGood** [comment]
@araq:

**Since the documentation for `=trace` would seem to be related to this issue, I will comment on that documentation here rather than clutter up somewhere else:**

**Your new documentation on `=trace` is clear enough but you might want to add something** about the following:

1. It isn't clearly mentioned that **`=destroy` and `=trace`** (which second type of "hook" is only of any use with `--gc:orc`) **are related** in that if the `=destroy` proc calls a field that is a reference to the same object before the `=trace` is defined, a `=trace` is automatically generated by the compiler for that object and vice versa - a `=trace` proc that calls a `=trace` on that same object automatically generates a `=destroy` unless `=destroy` has already been defined.  The solution for this mutual declaration is to forward declare the second defined one before the first defined one if there is a requirement to use both.
2. **The example for the use of a custom `=trace` isn't very clear on why one would do that.**  It is clear enough that one can extend the Orc tracing using it, and it is stated/implied that one would do that if using fields or pointer fields that could contain cyclic references.  The only purpose I can see of using a custom version to do this would be to change the order of the field traces from the default of top to bottom if that made the tracing more efficient;  Tracing a pointer to an object by dereferencing the pointer doesn't really make any sense as that pointer may as well be a `ref` in the first place (unless there is a lot of overhead using `ref`'s as compared to `ptr`'s that I don't know about.  The actual deallocation of pointers would still need to be done in the custom`=destructor` anyway.

I think users of these facilities would appreciate just a little more detail on these two points.

I would PR the Destructors page myself, but I'm afraid I might be missing some details and get it wrong.

**Comment by Araq** [comment]
> are related in that if the =destroy proc calls a field that is a reference to the same object before the =trace is defined, a =trace is automatically generated by the compiler for that object and vice versa - a =trace proc that calls a =trace on that same object automatically generates a =destroy unless =destroy has already been defined. The solution for this mutual declaration is to forward declare the second defined one before the first defined one if there is a requirement to use both.

That sounds like an issue worth fixing rather than documenting. No idea how hard that would be though.

> The example for the use of a custom =trace isn't very clear on why one would do that. 

You need to do it for custom containers or else these cannot be used for within cyclic structures in the sense that they are used to actually form the cycle. As a rule of thumb: write a `=trace` proc when you write a custom `destroy` proc that calls other `destroy` procs.

**Comment by GordonBGood** [comment]
@Araq:

> > are mutually related...
> 
> That sounds like an issue worth fixing rather than documenting. No idea how hard that would be though.

I have no idea how difficult either, but at least there is an easy work around in just forward declaring the second of the two proc's.  **Should I file an issue as a reminder of the behavior** and include the work around in the issue in case others have the same problem I did?

> > The example for the use of a custom `=trace` isn't very clear on why one would do that.
> 
> You need to do it for custom containers or else these cannot be used within cyclic structures in the sense that they are used to actually form the cycle. As a rule of thumb: write a `=trace` proc when you write a custom `=destroy` proc that calls other `=destroy` proc's.

Of course, although the `=trace` proc is only needed if `defined(gcOrc)`.  In fact, I did find a use for it in my new `std/channels` library as that library is based on buffered raw `ptr T`'s and in order to ensure that the `T`'s get destroyed properly in the case they are complex, I need to do a proper deep destruction in the `=destroy` proc, which since this library is intended to be usable with `--gc:orc` and possible cyclic references passed to the channel, the interior pointer buffered `T` types need to be traced.  It wasn't clear to me until I actually used it what the use is.

**So yes, when one needs a `=destroy` proc, one will almost always need a `=trace` proc if there is any possibility of cyclic references.**

**I assume that the `=trace` proc will just be ignored unless `--gc:orc` is being used?**

**EDITED:** Removed references to a regression as the issue was specific to use of  `--threads:on` with `--gc:arc` or `--gc:orc` and unrelated to this issue, which new issue has been reported as issue #18494.
